### ISSUE TITLE
Behaviour change with .whereIn non array|callback|builder argument handling

### ISSUE THREAD

**Issue by hpeinar** [tgriesser/knex#3201]
# Environment

Knex version: Bug introduced in 0.14.5, present in current latest from npm 0.16.5
Database: Tested with dialect mysql2

# Bug

There was a silent change in how **non** array|callback|builder arguments are handled for .whereIn in version 0.14.5 and it now produces invalid SQL.
Probably introduced with a fix for #1390
I didn't find any actual documentation of how this should or is handled in knex.js but it was very convenient to use as there was no worry if the input was one ID or an array of IDs.
Probably should also be mentioned in upgrading instructions as this change is breaking old code.

Knex version 0.14.4 (also expected results):
```javascript
knex('table').whereIn('field', 'value').toSQL().sql
```

```SQL
select * from `table` where `field` in (?)
```

Knex version 0.14.5:
_Note the missing parentheses_
```javascript
knex('table').whereIn('field', 'value').toSQL().sql
```

```SQL
select * from `table` where `field` in ?
```

Knex version 0.16.5 (current version from @latest npm install):
```javascript
knex('table').whereIn('field', 'value').toSQL().sql
```

```SQL
select * from `table` where `field` in ?
```

Two last SQL generated are invalid and result in ER_PARSE_ERROR (tested in 5.7 MySQL).

# Test
I didn't actually get the unit tests to run but something like this should be enough to test this case.
```javascript
it('where ins with non array|callback|builder argument', function() {
  testsql(
    qb()
      .select('*')
      .from('users')
      .whereIn('id', 1),
    {
      mysql: {
        sql: 'select * from `users` where `id` in (?)',
        bindings: [1],
      },
      mssql: {
        sql: 'select * from [users] where [id] in (?)',
        bindings: [1],
      },
      pg: {
        sql: 'select * from "users" where "id" in (?)',
        bindings: [1],
      },
      'pg-redshift': {
        sql: 'select * from "users" where "id" in (?)',
        bindings: [1],
      },
    }
  );
});
```
# Runkit
https://runkit.com/embed/a8re7gry28cx

**Comment by wubzz** [comment]
Probably a miss when it was rewritten. I find the syntax awful to be honest, but since it's documented as accepting single column I guess this really is a bug.

**Comment by hpeinar** [comment]
I'm not sure it is documented as such? I looked at the current documentation from knexjs.org and it states:

> whereIn — .whereIn(column|columns, array|callback|builder) / .orWhereIn

it doesn't say anything about \~mixed\~ parameters. 

Is this documented under something else? Or am I missing something.

**Comment by wubzz** [comment]
Ah, my bad. Read documentation all wrong. So it isn't documented.

Personally I wouldn't want to support single value like this when the operator is _always_ a list/array. But no strong opinion. Leaving bug label for now.

**Comment by JakeHamix** [comment]
We've just upgraded to a newer version of Node and Knex along with it. Can confirm that the issue originates from the reported version **0.14.5** and [#1390](https://github.com/tgriesser/knex/pull/1390/commits/50f2752b4927eda1279fbc130667cfed6c2fad7f). 

It caused us a lot of potential issues, because we rely on this behavior in many places. I ended up forking the repository and making a simple fix to repliate the wanted behavior. Gonna take a while until that code is properly tested and rolled out, will make a pull request when that happens, so far seems without any side effects.

**Comment by elhigu** [comment]
Throwing an error would sound like a correct functionality to me. It would catch the cases faster, where one passes something to the builder that was not meant to be passed (single value instead of an array / builder / raw).

**Comment by JakeHamix** [comment]
Either way, the behavior is inconsistent right now. We allow our application to be run above Oracle or Mssql databases, but this issue is not present with Oracle - not sure about others DBs.

**Comment by elhigu** [comment]
> Either way, the behavior is inconsistent right now. We allow our application to be run above Oracle or Mssql databases, but this issue is not present with Oracle - not sure about others DBs.

Yep, you should not pass single values to `whereIn()`. In documentation that is undefined behaviour.  When you pass an array (like it is documented) it is consistent.

**Comment by JakeHamix** [comment]
Fair enough, you make a valid point. In that case throwing an error sounds indeed like the desired behavior.

**Comment by pavanmehta91** [comment]
We're stuck with `0.13.0` due to this issue.
We don't have tests written for our application. 
Sorely missing extending QueryBuilder feature which came after `0.19`

**Comment by kibertoad** [comment]
@pavanmehta91 Why can't you just pass array always?

**Comment by pavanmehta91** [comment]
We definitely should do that. But there's a lot of code written earlier. There are 800 odd instances where we've used whereIn and nobody in the team is confident that all the places receive array in the argument.

**Comment by elhigu** [comment]
> We don't have tests written for our application.

Maybe it is good time to write tests if you are going to be maintaining that code (like updating libraries to non backwards compatible versions).

**Comment by pavanmehta91** [comment]
Yes true that.

**Comment by mercmobily** [comment]
Quick summary...

This is a behavioural change from 0.14.5 that we should address either by documenting or by enforcing a clearer error. `whereIn(column, value)` was never documented to accept a scalar; only array|callback|builder are supported. Some dialects happened to coerce a single value into `in (?)` previously, others now emit invalid SQL (`in ?`).

Code context: builder/formatter path does not normalise scalar to array for `whereIn` and should not silently guess. Cross‑dialect consistency matters here.

One possible resolution:

- Treat a scalar `value` as invalid usage and throw a clear error, recommending `whereIn('id', [value])` instead. That removes ambiguous behaviour and is consistent across dialects.
- Add tests to enforce this and update upgrading notes.

In the meantime, always pass an array to `whereIn`.
